server/sonar-server/src/main/java/org/sonar/server/permission/index/AuthorizationIndexer.java

```diff
@@ -0,0 +1,202 @@
+import org.elasticsearch.action.search.SearchResponse;
+import org.picocontainer.Startable;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.server.component.es.ProjectMeasuresIndexDefinition;
+import org.sonar.server.es.BulkIndexer;
+import org.sonar.server.es.EsClient;
+import org.sonar.server.es.EsUtils;
+import org.sonar.server.issue.index.IssueIndexDefinition;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Collections.singletonList;
+import static org.elasticsearch.index.query.QueryBuilders.matchAllQuery;
+
+/**
+ * Manages the synchronization of index issues/authorization with authorization settings defined in database :
+ * <ul>
+ *   <li>index the projects with recent permission changes</li>
+ *   <li>delete project orphans from index</li>
+ * </ul>
+ */
+public class AuthorizationIndexer implements Startable {
+
+  private static final int MAX_BATCH_SIZE = 1000;
+
+  private static final String BULK_ERROR_MESSAGE = "Fail to index authorization";
+
+  private final ThreadPoolExecutor executor;
+  private final DbClient dbClient;
+  private final EsClient esClient;
+
+  public AuthorizationIndexer(DbClient dbClient, EsClient esClient) {
+    this.executor = new ThreadPoolExecutor(0, 1, 0L, TimeUnit.SECONDS, new LinkedBlockingQueue<>());
+    this.dbClient = dbClient;
+    this.esClient = esClient;
+  }
+
+  /**
+   * Index issues authorization and project measures authorization indexes only when they are empty
+   */
+  public void indexAllIfEmpty() {
+    Future submit = executor.submit(() -> {
+      if (isIndexEmpty(IssueIndexDefinition.INDEX, IssueIndexDefinition.TYPE_AUTHORIZATION) ||
+        isIndexEmpty(ProjectMeasuresIndexDefinition.TYPE_PROJECT_MEASURES, ProjectMeasuresIndexDefinition.TYPE_AUTHORIZATION)) {
+        truncate(IssueIndexDefinition.INDEX, IssueIndexDefinition.TYPE_AUTHORIZATION);
+        truncate(ProjectMeasuresIndexDefinition.TYPE_PROJECT_MEASURES, ProjectMeasuresIndexDefinition.TYPE_AUTHORIZATION);
+        try (DbSession dbSession = dbClient.openSession(false)) {
+          index(new AuthorizationDao().selectAll(dbClient, dbSession));
+        }
+      }
+    });
+    try {
+      Uninterruptibles.getUninterruptibly(submit);
+    } catch (ExecutionException e) {
+      Throwables.propagate(e);
+    }
+  }
+
+  private boolean isIndexEmpty(String index, String type) {
+    SearchResponse issuesAuthorizationResponse = esClient.prepareSearch(index).setTypes(type).setSize(0).get();
+    return issuesAuthorizationResponse.getHits().getTotalHits() == 0;
+  }
+
+  private void truncate(String index, String type) {
+    BulkIndexer.delete(esClient, index, esClient.prepareSearch(index).setTypes(type).setQuery(matchAllQuery()));
+  }
+
+  public void index(List<String> projectUuids) {
+    checkArgument(!projectUuids.isEmpty(), "ProjectUuids cannot be empty");
+    try (DbSession dbSession = dbClient.openSession(false)) {
```

Comment:
The dbSession should not be opened here but passed as a parameter. Indeed a session is already open in PermissionService.